src/main/java/org/prebid/server/bidder/readpeak/ReadPeakBidder.java

```diff
@@ -0,0 +1,224 @@
+                .build();
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(BidderCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return extractBids(bidResponse);
+        } catch (DecodeException | PreBidException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private Result<List<BidderBid>> extractBids(BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return Result.withValues(Collections.emptyList());
+        }
+        return bidsFromResponse(bidResponse);
+    }
+
+    private Result<List<BidderBid>> bidsFromResponse(BidResponse bidResponse) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<BidderBid> validBids = new ArrayList<>();
+
+        for (SeatBid seatBid : bidResponse.getSeatbid()) {
+            if (seatBid != null) {
+                for (Bid bid : seatBid.getBid()) {
+                    if (bid != null) {
+                        try {
+                            final BidderBid bidderBid = makeBid(bid, bidResponse);
+                            validBids.add(bidderBid);
+                        } catch (PreBidException e) {
+                            errors.add(BidderError.badInput(e.getMessage()));
+                        }
+                    }
+                }
+            }
+        }
+
+        if (!errors.isEmpty()) {
+            return Result.withErrors(errors);
+        }
+
+        return Result.withValues(validBids);
+    }
+
+    private BidderBid makeBid(Bid bid, BidResponse bidResponse) {
+        final Bid resolvedBid = resolveMacros(bid);
+        final BidType bidType = getBidType(bid);
+        final Bid updatedBid = addBidMeta(resolvedBid);
+        return BidderBid.of(updatedBid, bidType, bidResponse.getCur());
+    }
+
+    private Bid resolveMacros(Bid bid) {
+        final BigDecimal price = bid.getPrice();
+        final String priceAsString = price != null ? price.toPlainString() : "0";
+
+        return bid.toBuilder()
+                .nurl(StringUtils.replace(bid.getNurl(), PRICE_MACRO, priceAsString))
+                .adm(StringUtils.replace(bid.getAdm(), PRICE_MACRO, priceAsString))
+                .burl(StringUtils.replace(bid.getBurl(), PRICE_MACRO, priceAsString))
+                .build();
+    }
+
+    private BidType getBidType(Bid bid) {
+        final Integer markupType = ObjectUtils.defaultIfNull(bid.getMtype(), 0);
+
+        return switch (markupType) {
+            case 1 -> BidType.banner;
+            case 4 -> BidType.xNative;
+            default -> throw new PreBidException(
+                    "Unable to fetch mediaType " + bid.getMtype() + " in multi-format: " + bid.getImpid());
+        };
```

Comment:
throwing exception leads to failing on the very first invalid bid, but the error should be just collected and other bids should be processed